src/template.js

```diff
@@ -139,99 +170,85 @@ var RegisteredChara = CreateClass({
-            this.setState({charaData: charadata})
+            this.setState({charaData: charadata});
         });
     },
     clickedTemplate: function (e) {
         this.props.onClick(this.state.charaData[e.target.getAttribute("id")]);
     },
     handleEvent: function (key, e) {
-        var newState = this.state;
-        newState[key] = e.target.value;
-        this.setState(newState)
+        this.setState({[key]: e.target.value});
     },
     openSendRequest: function (e) {
-        this.setState({openSendRequest: true})
+        this.setState({openSendRequest: true});
     },
     closeSendRequest: function (e) {
-        this.setState({openSendRequest: false})
+        this.setState({openSendRequest: false});
     },
     render: function () {
-        var clickedTemplate = this.clickedTemplate;
-        var filterText = this.state.filterText;
-        var filterElement = this.state.filterElement;
-        var charaData = this.state.charaData;
-        var limit = this.state.limit;
-        var displayed_count = 0;
-        var locale = this.props.locale;
+        const locale = this.props.locale;
+        const {
+            clickedTemplate,
+            filterText,
+            filterElement,
+            filterSex,
+            filterType,
+            filterRace,
+            filterFav,
+            charaData,
+            limit,
+        } = this.state;
+
+        const charaTemplateHeader = <>
+            <span>検索:</span>
+            <FormControl type="text" placeholder={intl.translate("キャラ名", locale)} value={this.state.filterText}
+                         onChange={this.handleEvent.bind(this, "filterText")}/>
+            <FormControl componentClass="select" value={this.state.filterElement}
+                         onChange={this.handleEvent.bind(this, "filterElement")}>{selector[locale].filterElements}</FormControl>
+            <FormControl componentClass="select" value={this.state.filterSex}
+                         onChange={this.handleEvent.bind(this, "filterSex")}>{selector[locale].filterSexes}</FormControl>
+            <FormControl componentClass="select" value={this.state.filterType}
+                         onChange={this.handleEvent.bind(this, "filterType")}>{selector[locale].filterTypes}</FormControl>
+            <FormControl componentClass="select" value={this.state.filterRace}
+                         onChange={this.handleEvent.bind(this, "filterRace")}>{selector[locale].filterRaces}</FormControl>
+            <FormControl componentClass="select" value={this.state.filterFav}
+                         onChange={this.handleEvent.bind(this, "filterFav")}>{selector[locale].filterFavs}</FormControl>
+            </>;
+
+        const filterFunc = _generateCharaFilterFunc(this.state, locale);
+
+        const mapFunc = ([key,val]) =>
+            <div className="onechara" key={key}>
+                <p>{val[locale]}</p><br/>
+                <Image rounded onClick={clickedTemplate} id={key}
```

Comment:
clickedTemplate will be always undefined. It needs to be set during render or referenced in initialState or replaced with this.clickedTemplate

Reply:
thank u for report, seem I did not do the acceptance test.
I see that context of the closure. is it because of JSX component context?

- key, val ... are arguments ok
- clickedTemplate ... outer scope variable (different context in run-time)
- locale ... also undefined ?

I will try this.clickedTemplate as quick solution, and also fix "locale" if needed.
I think ideally those closure should be designed as testable function. (future issue)

Reply:
my misunderstood, it was not context issue.
at the head of function, I was trying to import clickTemplate local variable from "this.state"
locale was ok, const clickedTemplate was undefined